pkg/controllers/dashboard/plugin/fscache.go

```diff
@@ -0,0 +1,253 @@
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Write(data)
+
+	return nil
+}
+
+// Delete takes in a plugin's name and version, and deletes its entry in the filesystem cache
+func (c FSCache) Delete(name, version string) error {
+	p, err := filepathsecure.SecureJoin(FSCacheRootDir, name)
+	if err != nil {
+		return err
+	}
+	err = osRemoveAll(p)
+	if err != nil {
+		err = fmt.Errorf("failed to delete entry [Name: %s Version: %s] from filesystem cache: %s", name, version, err.Error())
+		return err
+	}
+	logrus.Debugf("deleted plugin entry from cache [Name: %s Version: %s]", name, version)
+
+	return nil
+}
+
+// isCached takes in the name and version of a plugin and returns true if
+// it is cached (entry exists and files were fetched), returns false otherwise
+func (c FSCache) isCached(name, version string) (bool, error) {
+	path, err := filepathsecure.SecureJoin(FSCacheRootDir, filepath.Join(name, version))
+	if err != nil {
+		return false, err
+	}
+	_, err = osStat(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+		isEmpty, err := isDirEmpty(path)
+		if err != nil {
+			return false, err
+		}
+		if !isEmpty {
+			return true, nil
+		}
+
+		return false, err
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		isEmpty, err := isDirEmpty(path)
+		if err != nil {
+			return false, err
+		} else if !isEmpty {
+			return true, nil
+		}
+	} else if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+func fsCacheFilepathGlob(pattern string) ([]string, error) {
+	files, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, err
+	}
+	logrus.Debugf("files matching glob pattern [%s] found in filesystem cache: %+v", pattern, files)
+
+	return files, nil
+}
+
+// getVersionFromPackageJSON takes in a URL for a plugin's package.json, reads it, and returns a Semver object of the version contained in the file
+func getVersionFromPackageJSON(packageJSONURL string) (*semver.Version, error) {
+	data, err := readFile(packageJSONURL)
+	if err != nil {
+		return nil, err
+	}
+	var packageJSON PackageJSON
+	err = json.Unmarshal(data, &packageJSON)
+	if err != nil {
+		return nil, err
+	}
+	version, err := semver.NewVersion(packageJSON.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	return version, nil
+}
+
+// readFilesTxt takes in a URL for a plugin's files.txt, reads it, and returns a slice of the file paths contained in the file
+func readFilesTxt(filesTxtURL string) ([]string, error) {
+	data, err := readFile(filesTxtURL)
+	if err != nil {
+		return nil, err
+	}
+	files := strings.Split(string(data), "\n")
+
+	return files, nil
+}
+
+// readFile reads the file from the given URL and returns the data
+func readFile(URL string) ([]byte, error) {
+	logrus.Debugf("fetching file [%s]", URL)
+	resp, err := http.Get(URL)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	data, err := io.ReadAll(resp.Body)
```

Comment:
What is the expected file size here? Let's add a limit reader and ensure there is a max amount of data that would be loaded into memory.

Reply:
I don't think there's an expected size, but they shouldn't be too big since it's usually `.js` files and such

Reply:
Resolving this item as per off-line discussion on limiting the amount of resources loaded into memory.